File: src/closing.zig
const std = @import("std");
const db_mod = @import("db.zig");
const models = @import("models.zig");
const report = @import("report.zig");
const archive = @import("archive.zig");
const attachment_mod = @import("attachment.zig");

pub const CloseType = enum {
    month,
    year,
};

pub fn closePeriod(allocator: std.mem.Allocator, database: *db_mod.Db, basePath: []const u8, closeType: CloseType, period: []const u8) !void {
    const typeStr: [:0]const u8 = switch (closeType) {
        .month => "month",
        .year => "year",
    };

    if (try database.isClosed(period, typeStr)) {
        std.log.err("Period {s} ({s}) is already closed", .{ period, typeStr });
        return error.AlreadyClosed;
    }

    const invoices = try queryInvoicesForPeriod(allocator, database, closeType, period);
    defer {
        for (invoices) |inv| {
            allocator.free(inv.number);
            allocator.free(inv.date);
            allocator.free(inv.type);
            allocator.free(inv.item_name);
            allocator.free(inv.seller_name);
            allocator.free(inv.seller_tax_id);
            allocator.free(inv.buyer_name);
            allocator.free(inv.buyer_tax_id);
            allocator.free(inv.category);
            allocator.free(inv.remark);
            allocator.free(inv.created_at);
            allocator.free(inv.updated_at);
        }
        allocator.free(invoices);
    }

    if (invoices.len == 0) {
        std.log.err("No invoices found for period {s}", .{period});
        return error.NoInvoices;
    }

    var detailPathBuf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    var summaryPathBuf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const detailPath = try std.fmt.bufPrintZ(&detailPathBuf, "{s}/.invoice/明细表_{s}.xlsx", .{ basePath, period });
    const summaryPath = try std.fmt.bufPrintZ(&summaryPathBuf, "{s}/.invoice/汇总表_{s}.xlsx", .{ basePath, period });

    try report.generateDetailReport(invoices, detailPath);

    const summaryEntries = try report.computeSummary(allocator, invoices);
    defer allocator.free(summaryEntries);
    try report.generateSummaryReport(summaryEntries, summaryPath);

    var attList = std.ArrayList(archive.AttachmentEntry).init(allocator);
    defer attList.deinit();

    for (invoices) |inv| {
        const atts = attachment_mod.getAttachmentsForInvoice(database, allocator, inv.id) catch &[_]models.Attachment{};
        defer allocator.free(atts);
        for (atts) |att| {
            try attList.append(.{
                .invoice_number = inv.number,
                .filepath = att.filepath,
                .filename = att.filename,
            });
        }
    }

    var archivePathBuf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const archivePath = try std.fmt.bufPrintZ(&archivePathBuf, "{s}/.invoice/close_{s}.zip", .{ basePath, period });

    try archive.createArchive(allocator, detailPath, summaryPath, attList.items, archivePath);

    const insertSql = "INSERT INTO closings (type, period, archive_path) VALUES (?, ?, ?)";
    const insertStmt = try database.prepare(insertSql);
    defer insertStmt.deinit();

    var typeBuf: [64]u8 = undefined;
    var periodBuf: [32]u8 = undefined;
    var archiveBuf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const typeZ = try std.fmt.bufPrintZ(&typeBuf, "{s}", .{typeStr});
    const periodZ = try std.fmt.bufPrintZ(&periodBuf, "{s}", .{period});
    const archiveZ = try std.fmt.bufPrintZ(&archiveBuf, ".invoice/close_{s}.zip", .{period});

    try insertStmt.bindText(1, typeZ);
    try insertStmt.bindText(2, periodZ);
    try insertStmt.bindText(3, archiveZ);

    const rowDone = try insertStmt.step();
    if (rowDone) return error.UnexpectedRow;
}

pub fn checkPeriodClosed(database: *db_mod.Db, date: []const u8) !bool {
    if (date.len >= 7) {
        const month = date[0..7];
        if (try database.isClosed(month, "month")) return true;
    }
    if (date.len >= 4) {
        const year = date[0..4];
        if (try database.isClosed(year, "year")) return true;
    }
    return false;
}

pub fn queryInvoicesForPeriod(allocator: std.mem.Allocator, database: *db_mod.Db, closeType: CloseType, period: []const u8) ![]models.Invoice {
    _ = closeType;
    var list = std.ArrayList(models.Invoice).init(allocator);
    errdefer list.deinit();

    const sql = "SELECT id, number, date, type, item_name, amount, tax_rate, tax, total, seller_name, seller_tax_id, buyer_name, buyer_tax_id, category, remark, created_at, updated_at FROM invoices WHERE date LIKE ? ORDER BY date, id";
    const stmt = try database.prepare(sql);
    defer stmt.deinit();

    var patternBuf: [64]u8 = undefined;
    const pattern = try std.fmt.bufPrintZ(&patternBuf, "{s}-%", .{period});
    try stmt.bindText(1, pattern);

    while (try stmt.step()) {
        const inv = try models.Invoice.fromRowAlloc(stmt, allocator);
        try list.append(inv);
    }

    return try list.toOwnedSlice();
}

File: src/closing.rs
use rusqlite::params;

use crate::archive;
use crate::db;
use crate::models;
use crate::report;

#[derive(Debug, Clone, Copy)]
pub enum CloseType {
    Month,
    Year,
}

impl CloseType {
    pub fn as_str(&self) -> &'static str {
        match self {
            CloseType::Month => "month",
            CloseType::Year => "year",
        }
    }
}

pub fn close_period(
    conn: &rusqlite::Connection,
    close_type: CloseType,
    period: &str,
) -> Result<(), Box<dyn std::error::Error>> {
    let type_str = close_type.as_str();

    if db::is_period_closed(conn, period, type_str)? {
        return Err(format!("Period {} ({}) is already closed", period, type_str).into());
    }

    let invoices = query_invoices_for_period(conn, close_type, period)?;
    if invoices.is_empty() {
        return Err(format!("No invoices found for period {}", period).into());
    }

    let cwd = std::env::current_dir()?;
    let detail_path = cwd.join(format!(".invoice/明细表_{}.xlsx", period));
    let summary_path = cwd.join(format!(".invoice/汇总表_{}.xlsx", period));

    report::generate_detail_report(&invoices, detail_path.to_str().unwrap_or(""))?;
    let summary_entries = report::compute_summary(&invoices);
    report::generate_summary_report(&summary_entries, summary_path.to_str().unwrap_or(""))?;

    let mut att_list: Vec<archive::AttachmentEntry> = Vec::new();
    for inv in &invoices {
        let atts = db::get_attachments_for_invoice(conn, inv.id)?;
        for att in atts {
            att_list.push(archive::AttachmentEntry {
                invoice_number: inv.number.clone(),
                filepath: att.filepath.clone(),
                filename: att.filename.clone(),
            });
        }
    }

    let archive_path = cwd.join(format!(".invoice/close_{}.zip", period));
    archive::create_archive(
        detail_path.to_str().unwrap_or(""),
        summary_path.to_str().unwrap_or(""),
        &att_list,
        archive_path.to_str().unwrap_or(""),
    )?;

    let db_archive_path = format!(".invoice/close_{}.zip", period);
    db::insert_closing(conn, type_str, period, &db_archive_path)?;

    Ok(())
}

pub fn check_invoice_closed(
    conn: &rusqlite::Connection,
    invoice_id: i64,
) -> Result<bool, Box<dyn std::error::Error>> {
    let date: Option<String> = conn
        .query_row(
            "SELECT date FROM invoices WHERE id = ?1",
            params![invoice_id],
            |row| row.get(0),
        )
        .ok();

    match date {
        Some(d) => check_period_closed(conn, &d),
        None => Ok(false),
    }
}

pub fn check_period_closed(
    conn: &rusqlite::Connection,
    date: &str,
) -> Result<bool, Box<dyn std::error::Error>> {
    if date.len() >= 7 {
        let month = &date[..7];
        if db::is_period_closed(conn, month, "month")? {
            return Ok(true);
        }
    }
    if date.len() >= 4 {
        let year = &date[..4];
        if db::is_period_closed(conn, year, "year")? {
            return Ok(true);
        }
    }
    Ok(false)
}

pub fn query_invoices_for_period(
    conn: &rusqlite::Connection,
    _close_type: CloseType,
    period: &str,
) -> Result<Vec<models::Invoice>, Box<dyn std::error::Error>> {
    let pattern = format!("{}-%", period);
    let mut stmt = conn.prepare(
        "SELECT id, number, date, type, item_name, amount, tax_rate, tax, total, seller_name, seller_tax_id, buyer_name, buyer_tax_id, category, remark, created_at, updated_at FROM invoices WHERE date LIKE ?1 ORDER BY date, id"
    )?;
    let invoices = stmt
        .query_map(params![pattern], |row| {
            Ok(models::Invoice {
                id: row.get(0)?,
                number: row.get(1)?,
                date: row.get(2)?,
                inv_type: row.get(3)?,
                item_name: row.get(4)?,
                amount: row.get(5)?,
                tax_rate: row.get(6)?,
                tax: row.get(7)?,
                total: row.get(8)?,
                seller_name: row.get(9)?,
                seller_tax_id: row.get(10)?,
                buyer_name: row.get(11)?,
                buyer_tax_id: row.get(12)?,
                category: row.get(13)?,
                remark: row.get(14)?,
                created_at: row.get(15)?,
                updated_at: row.get(16)?,
            })
        })?
        .collect::<Result<Vec<_>, _>>()?;
    Ok(invoices)
}

#[cfg(test)]
mod tests {
    #[test]
    fn test_close_type_as_str() {
        assert_eq!(super::CloseType::Month.as_str(), "month");
        assert_eq!(super::CloseType::Year.as_str(), "year");
    }
}
